kube-core/src/version.rs

```diff
@@ -0,0 +1,396 @@
+use std::{cmp::Reverse, convert::Infallible, str::FromStr};
+
+/// Version parser for Kubernetes version patterns
+///
+/// This type implements two orderings for sorting by:
+///
+/// - [`Version::priority`] for [Kubernetes/kubectl version priority](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority)
+/// - [`Version::generation`] for sorting strictly by version generation in a semver style
+///
+/// To get the api versions sorted by `kubectl` priority:
+///
+/// ```
+/// use kube_core::Version;
+/// use std::cmp::Reverse; // for DESCENDING sort
+/// let mut versions = vec![
+///     "v10beta3",
+///     "v2",
+///     "foo10",
+///     "v1",
+///     "v3beta1",
+///     "v11alpha2",
+///     "v11beta2",
+///     "v12alpha1",
+///     "foo1",
+///     "v10",
+/// ];
+/// versions.sort_by_cached_key(|v| Reverse(Version::parse(v).priority()));
+/// assert_eq!(versions, vec![
+///     "v10",
+///     "v2",
+///     "v1",
+///     "v11beta2",
+///     "v10beta3",
+///     "v3beta1",
+///     "v12alpha1",
+///     "v11alpha2",
+///     "foo1",
+///     "foo10",
+/// ]);
+/// ```
+///
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum Version {
+    /// A major/GA release
+    ///
+    /// Always considered higher priority than a beta release.
+    Stable(u32),
+
+    /// A beta release for a specific major version
+    ///
+    /// Always considered higher priority than an alpha release.
+    Beta(u32, Option<u32>),
+
+    /// An alpha release for a specific major version
+    ///
+    /// Always considered higher priority than a nonconformant version
+    Alpha(u32, Option<u32>),
+    /// An non-conformant api string
+    ///
+    /// CRDs and APIServices can use arbitrary strings as versions.
+    Nonconformant(String),
+}
+// NB:
+
+impl Version {
+    fn try_parse(v: &str) -> Option<Version> {
+        let v = v.strip_prefix('v')?;
+        let major_chars = v.chars().take_while(|ch| ch.is_ascii_digit()).count();
+        let major = &v[..major_chars];
```

Comment:
FWIW, this assumes that UTF-8 byte length == Unicode code point count. This isn't an issue here (we explicitly only care about ASCII characters), but it's still worth keeping straight.. PR upcoming.

Reply:
https://github.com/kube-rs/kube-rs/pull/772